dbms/src/Operators/Operator.h

```diff
@@ -0,0 +1,89 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Core/Block.h>
+
+#include <memory>
+
+namespace DB
+{
+enum class OperatorStatus
+{
+    // spilling status
+    SPILLING,
+    // waiting status
+    WAITING,
+    SKIP,
+    // running status
+    PASS,
+    NO_OUTPUT,
+    MORE_INPUT,
+    // finish status
+    FINISHED,
+    CANCELLED,
+};
+
+class SpillOp
+{
+public:
+    virtual ~SpillOp() = default;
+
+    virtual OperatorStatus spill() { throw Exception("Unsupport"); }
+};
+
+// TODO support profile info like `BlockStreamProfileInfo`.
+
+class Source : public SpillOp
+{
+public:
+    virtual ~Source() = default;
+
+    virtual OperatorStatus read(Block & block) = 0;
+
+    virtual Block readHeader() = 0;
+
+    virtual OperatorStatus await() { return OperatorStatus::PASS; }
+};
+using SourcePtr = std::unique_ptr<Source>;
+
+class Transform : public SpillOp
+{
+public:
+    virtual ~Transform() = default;
+
+    // call fetchBlock first, and then call transform.
+    virtual OperatorStatus fetchBlock(Block &) { return OperatorStatus::NO_OUTPUT; }
+    virtual OperatorStatus transform(Block & block) = 0;
```

Comment:
Since `transform` will change the `block` inplace, if the transform will expand the block, we have to return a huge block after the transform ?

Reply:
Yes, so we need to ensure that the operator will not return too large a block at a time...

Reply:
Maybe the issues like https://github.com/pingcap/tiflash/issues/6579 will help a lot.